src/main/java/io/vertx/core/impl/VertxImpl.java

```diff
@@ -581,14 +581,14 @@ public synchronized void close(Handler<AsyncResult<Void>> completionHandler) {
         }
         haPromise.future().onComplete(ar2 -> {
           addressResolver.close(ar3 -> {
-            Promise<Void> ebClose = getOrCreateContext().promise();
-            eventBus.close(ebClose);
-            ebClose.future().onComplete(ar4 -> {
-              closeClusterManager(ar5 -> {
-                // Copy set to prevent ConcurrentModificationException
-                deleteCacheDirAndShutdown(completionHandler);
+              closeClusterManager(ar4 -> {
+                Promise<Void> ebClose = getOrCreateContext().promise();
+                eventBus.close(ebClose);
+                ebClose.future().onComplete(ar5 -> {
+                  // Copy set to prevent ConcurrentModificationException
+                  deleteCacheDirAndShutdown(completionHandler);
+                });
```

Comment:
currently ClusterEventBus is closed first which closes the NetServer and then ClusterManager which removes all MessageConsumer subscriptions.

maybe this is super naive of me but can this be just turned around?

**NOTE**: WIP commit

Reply:
I think it makes sense if we are sure that the parts of the clustered event bus used by the event bus will work correctly when the cluster manager is closed, i.e the closed node selector will return a failure, currently it does not seem to.

@tsegismont WDYT ?

Reply:
the interesting question is the order of shutdown... is there any "lifecycle" documented or implemented?

IMHO the ideal sequence would be:
1. discovery unpublish self
2. --optional delay--
3. shutoff servers / external consumers (httpserver, netserver, custer message consumers)
4. shutoff timers
5. verticle stop()
6. local message consumers
7. clients (httpclient, netclient, mongoclient, ....)
8. clustered maps / locks / counter